src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeBlob.java

```diff
@@ -86,6 +115,37 @@ public void update(Observable o, Object data) {
     });
   }
 
+  public static Class<?> getClassFor(Address addr) {
+      CodeBlob cb = new CodeBlob(addr);
+      if (cb.isNMethod()) {
+          return NMethod.class;
+      } else if (cb.isBufferBlob()) {
+          return BufferBlob.class;
+      } else if (cb.isAdapterBlob()) {
+          return AdapterBlob.class;
+      } else if (cb.isVtableBlob()) {
+          return VtableBlob.class;
+      } else if (cb.isMHAdapterBlob()) {
+          return MethodHandlesAdapterBlob.class;
+      } else if (cb.isRuntimeStub()) {
+          return RuntimeStub.class;
+      } else if (cb.isDeoptimizationBlob()) {
+          return DeoptimizationBlob.class;
+      } else if (cb.isSafepointBlob()) {
+          return SafepointBlob.class;
+      } else if (cb.isUpcallStub()) {
+          return UpcallStub.class;
+      }
+      if (VM.getVM().isServerCompiler()) {
+        if (cb.isExceptionBlob()) {
+            return ExceptionBlob.class;
+        } else if (cb.isUncommonTrapBlob()) {
+            return UncommonTrapBlob.class;
+        }
+      }
+      return null;
```

Comment:
Should this be an assert?

Reply:
I don't think we need it - the caller `CodeCache.createCodeBlobWrapper()` will throw `RuntimeException` when `null` is returned.

Reply:
I guess my real question is whether or not it can be considered normal behavior to return null. It seems it should never happen, which is why I was suggesting an assert.

Reply:
With your suggested `wrapperClasses[]` we will get OOB exception. No need separate assert.